Docs/doc 95/doc 110/technical term validation result (#4122)
great-expectations/great_expectations: docs/terms/validation_result.md

import TechnicalTag from '../term_tags/_tag.mdx';
import CreateHeader from '/docs/images/universal_map/_um_create_header.mdx';
import ValidateHeader from '/docs/images/universal_map/_um_validate_header.mdx';

<UniversalMap setup='inactive' connect='inactive' create='active' validate='active'/> 

## Overview

### Definition

A Validation Result is an object generated when data is validated against an <TechnicalTag relative="../" tag="expectation" text="Expectation" /> or <TechnicalTag relative="../" tag="expectation_suite" text="Expectation Suite" />.

### Features and promises

Validation Results can be saved for every time that a <TechnicalTag relative="../" tag="checkpoint" text="Checkpoint" /> <TechnicalTag relative="../" tag="validation" text="Validates" /> data, permitting you to maintain records of your data's adherence to Expectations and track trends in the quality of the validated data.  Validation Results are also among the information passed to <TechnicalTag relative="../" tag="action" text="Actions" /> when a Checkpoint is run.  This allows them to be used for any purpose that an Action can be created to handle. 

### Relationship to other objects

> Validation Results are generated inside a Checkpoint when a <TechnicalTag relative="../" tag="validator" text="Validator" /> runs an Expectation Suite against its paired <TechnicalTag relative="../" tag="batch_request" text="Batch Request" />.  The Validation Result will be passed to the Checkpoint's `action_list` where it will be available to any Actions that may need it.  The `StoreValidationResultAction` subclass of `ValidationAction` will save Validation Results to the <TechnicalTag relative="../" tag="validation_result_store" text="Validation Result Store" /> if it is present in a Checkpoint's `action_list`.

Reviewer: Minor suggestion, feel free to ignore.

```suggestion
Validation Results are generated by a Checkpoint when a <TechnicalTag relative="../" tag="validator" text="Validator" /> runs an Expectation Suite against its paired <TechnicalTag relative="../" tag="batch_request" text="Batch Request" />.  The Validation Result will be passed to the Checkpoint's `action_list` where it will be available to any Actions that may need it.  The `StoreValidationResultAction` subclass of `ValidationAction` will save Validation Results to the <TechnicalTag relative="../" tag="validation_result_store" text="Validation Result Store" /> if it is present in a Checkpoint's `action_list`.
```

## Use cases

<CreateHeader/>

> A Validation Result is generated when you use [the interactive workflow to create Expectations](../guides/expectations/how_to_create_and_edit_expectations_with_instant_feedback_from_a_sample_batch_of_data.md) and the newly created Expectation is run against the sample <TechnicalTag relative="../" tag="batch" text="Batch" /> of data you are working with.  These individual Validation Results will be created as an `ExpectationValidationResult` instance.

Reviewer: ```suggestion
A Validation Result is generated when you use [the interactive workflow to create Expectations](../guides/expectations/how_to_create_and_edit_expectations_with_instant_feedback_from_a_sample_batch_of_data.md) and the newly created Expectation is run against the sample <TechnicalTag relative="../" tag="batch" text="Batch" /> of data you are working with.  These individual Validation Results will be created as `ExpectationValidationResult` instances.
```

<ValidateHeader/>

> Validation Results are generated when you run Checkpoints to Validate data.  Because Checkpoints operate on Expectation Suites rather than individual Expectations, the Validation Results generated by Checkpoints will be created as `ExpectationSuiteValidationResult` instances. The difference between an `ExpectationValidationResult` instance and an `ExpectationSuiteValidationResult` instance is that the `ExpectationSuiteValidationResult` will contain additional information pertaining to the Expectation Suite that was Validated, and will contain the data of an `ExpectationValidationResult` for each Expectation in the Expectation Suite.  These `ExpectationValidationResult` instances will be in a list accessed by the `results` attribute of the `ExpectationSuiteValidationResult` instance.

Reviewer: I think there might be a more straightforward way of saying this - basically an `ExpectationSuiteValidationResult` is to an `ExpectationValidationResult` as an `ExpectationSuite` is to an `Expectation`. One contains one more of the other (though technically a suite can have 0 Expectations), plus some additional metadata.

Author: I added this clarification and moved the more specific (and harder to parse) explanation into a followup paragraph.

> The Validation Results generated as `ExpectationSuiteValidationResult` instances can be saved in JSON format in a Validation Result Store, and rendered as <TechnicalTag relative="../" tag="data_docs" text="Data Docs" />.  Both of these activities are handled by Actions in the Checkpoint's `action_list`.  In most cases, when you see "Validation Results" in Great Expectations, it will be in reference to the output generated as an `ExpectationSuiteValidationResult` when a Checkpoint is run.

Reviewer: Again, I'm being pendantic. "Into" could also work.

```suggestion
The Validation Results generated as `ExpectationSuiteValidationResult` instances can be saved in JSON format in a Validation Result Store, and rendered in <TechnicalTag relative="../" tag="data_docs" text="Data Docs" />.  Both of these activities are handled by Actions in the Checkpoint's `action_list`.  In most cases, when you see "Validation Results" in Great Expectations, it will be in reference to the output generated as an `ExpectationSuiteValidationResult` when a Checkpoint is run.
```

## Features

### Trends in data quality

> Saved Validation Results allow you to track trends in how well your data adheres to Expectations.  This can help you determine if failed Expectations are due to outliers in new data, or a more systemic problem.

Reviewer 2: This is super interesting and I dont think I thought about Validation results in this way before. I'm curious about how this works in OSS vs Cloud. 

Would the OSS version be something like: (sorry for the overly simple example) "if one of your Expectations has been failing every time for the last week, something is probably systematically wrong with your data"?

Reviewer: One example of this for me would be for illustrating data drift. Maybe you have `expect_column_mean_to_be_between(min_value=0, max_value=100)`. The first week, the mean might be 10, the next week 50, the next 100, the next 140.  Only the last one would fail, but the EVRs would tell a compelling story.

Overall, I love this use case, but we don't currently have a great way of actually doing this internally. Users who do this tend to parse the JSON and pass those results to a dashboard. I'm not sure whether or not that matters for the purposes of this doc.

Author: I was told that this was generally done manually, but it seems like a potentially important use of saved Validation Results.  I'm going to leave it in the doc for now -- and perhaps in the future we can come up with and implement a good way of doing this internally.

### Actions and documentation

Because Validation Results are passed to Actions in a Checkpoint's `action_list` you can create Actions that have behaviour conditional to the Validation Results.  Examples of this might be sending a Slack or email notification if the Validation fails, only launching a secondary process if the Validation Results report that Validation passed, or updating your Data Docs with specific information pulled from the Validation Results.

## API basics

### How to access

> Inside of Actions, Validation Results are a parameter that is passed to the Action when it is run.  The other place where you can access your Validation Results is in your Validation Result Store.  In both of these cases, you will be working with either an `ExpectationSuiteValidationResult` instance, or the serialization of such an instance into a JSON file.

Reviewer: The third option here is for the case where a user might be using a Database Store Backend. I think it's a less common use case, and would add some complexity to this definition, so we should just remove the "either" here.

```suggestion
Inside of Actions, Validation Results are a parameter that is passed to the Action when it is run.  The other place where you can access your Validation Results is in your Validation Result Store.  In both of these cases, you will be working with an `ExpectationSuiteValidationResult` instance, or the serialization of such an instance into a JSON file.
```

> Outside of Actions, you will generally only encounter Validation Results when validating a single expectation against some sample data in the interactive workflow for creating Expectations.  When an individual Validation Result is generated, it is generated as an `ExpectationValidationResult` instance.

Reviewer: ```suggestion
Outside of Actions and your Validation Result Store, you will generally only encounter Validation Results when validating a single expectation against some sample data in the interactive workflow for creating Expectations.  When an individual Validation Result is generated, it is generated as an `ExpectationValidationResult` instance.
```

A single Validation Result is created when an Expectation is created with the interactive workflow and the newly created Expectation validates itself against the sample data that is used in that process.  This is generated as an `ExpectationValidationResult` object.

A Validation Result is also created for each Expectation in an Expectation Suite when a Checkpoint is run.  These are generated as values in the `results` list of an `ExpectationSuiteValidationResult` object.

### Configuration

#### Expectation Validation Results

`ExpectationValidationResult` instances represent the Validation Result of a single Expectation.  If you are accessing the instance itself (rather than the serialized JSON of an instance found in your Validation Results Store) you will find that the instance provides information about the Expectation and the outcome of its Validation through the following attributes:

- **success:** A `true` or `false` indicator of whether the Expectation passed.
- **expectation_config:** The config used by the Expectation, including such things as type of Expectation and key word arguments that were passed to the Expectation.

> - **result:** The observed value generated when the Expectation was run.

Reviewer: ```suggestion
- **result:** The observed values generated when the Expectation was run.
```
